### PR TITLE
Add ATenOp and call aten::embedding and its Backward Op from ORT

### PR DESCRIPTION
**Pull request by iK1D**
The PR includes:
- Add ATenOp and its gradient to support calling ATen operators from ORT
- In the online torch extension, torch::jit::getAllOperatorsFor is used to get operator by name, and is executed using torch::jit::Stack
- Use aten::embedding as example, register com.microsoft::ATenOp for embedding for ORTModule

Ideally the ATen operator configs can be generated from text config automatically. Currently it's written as C++ code as example in this PR.

### CODE REVIEW

File: orttraining/orttraining/core/graph/training_op_defs.cc
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/training_op_defs.cc:2501]: I see... shape inference will start to fail starting from this point. 
One drawback of using ATenOps is that we will miss the shape inference on the forward graph, and it will result in ReduceSumTraining D2D memcpy.

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py
Comment (SherlockNoMad) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:114]: let's add a hard-coded flag to block this from using in the common exec path. 
We will move this load process to after getting the onnx exported graph 

self._use_aten_ops = False
if self._use_aten_ops: 
 ,,,,

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py
Comment (SherlockNoMad) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:567]: please switch to use _test_helpers.assert_values_are_close()

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py
Comment (SherlockNoMad) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:546]: how about using padding_idx ?

File: orttraining/orttraining/training_ops/cuda/aten_ops/aten_op.cc
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cuda/aten_ops/aten_op.cc:12]: I wonder do we need to set ExternalOutputs() here ?

As we are using aten kernels as a function, pytorch shouldn't need keep external reference of the output tensor?

File: orttraining/orttraining/training_ops/cpu/aten_ops/aten_op.cc
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/aten_ops/aten_op.cc:16]: Same here. Do we need to set ExternalOutputs()?

File: orttraining/orttraining/training_ops/cpu/aten_ops/aten_op.h
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/aten_ops/aten_op.h:20]: a comment here to indicate the size_t is the the input index for aten kernels?

File: cmake/onnxruntime_providers.cmake
Comment (SherlockNoMad) [inline cmake/onnxruntime_providers.cmake:212]: as we moved the dlpack to /core

in onnxruntime_python.cmake, do we still need to build with dlpack dependency? 
I see following line is still in python.cmake

set(DLPACK_INCLUDE_DIR ${PROJECT_SOURCE_DIR}/external/dlpack/include)

File: onnxruntime/python/onnxruntime_pybind_state.cc
Comment (SherlockNoMad) [inline onnxruntime/python/onnxruntime_pybind_state.cc:950]: When we have multiple ORTModule, this SetExecutorFunc will be called multiple times with same value?

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py
Comment (SherlockNoMad) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:114]: Also, how would this work when we have multiple-ortmodules? 
Will we compile the same extension code multiple times?

File: orttraining/orttraining/training_ops/cpu/aten_ops/aten_op_attr_parser.h
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/aten_ops/aten_op_attr_parser.h:49]: nit. if-elseif 

so that other checks can be skipped, if we find one macth

File: orttraining/orttraining/training_ops/cpu/aten_ops/aten_op.cc
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/aten_ops/aten_op.cc:45]: int_arguments_ is using int64_t. 
we are trying to parse by int?

File: orttraining/orttraining/training_ops/cpu/aten_ops/aten_op.cc
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/aten_ops/aten_op.cc:28]: nit... I think it worth writing an util function for check if aten ops is supported. Such check seems pretty common... 
Something lile
inline bool aten_ops::HasAtenOp(op_name) ? 
or 
inline bool aten_ops::TryGetAtenOp(op_name, op_config) ?

File: orttraining/orttraining/training_ops/cpu/aten_ops/aten_op.cc
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/aten_ops/aten_op.cc:70]: what happens when aten kernel throws exception? 
How does failures in aten kernel translate to non-OK Status in ORT?

File: orttraining/orttraining/training_ops/cpu/aten_ops/aten_op.cc
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/aten_ops/aten_op.cc:73]: Add a TODO here, need to support multiple output case .

File: orttraining/orttraining/training_ops/cpu/aten_ops/aten_op_config.h
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/aten_ops/aten_op_config.h:23]: a comment to make the mapping explicit. 

GO, I, O

Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/aten_ops/aten_op_config.h:23]: BackwardInputSource(Type/Kind)?

File: orttraining/orttraining/training_ops/cpu/aten_ops/aten_op_config.h
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/aten_ops/aten_op_config.h:18]: ForwardOutputTypeInfer(Kind/Type) ?

File: orttraining/orttraining/training_ops/cpu/aten_ops/aten_op_config.h
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/aten_ops/aten_op_config.h:47]: i for input? 
one pattern I see it to use prefix _
so _backward_op_name

File: orttraining/orttraining/training_ops/cpu/aten_ops/aten_op_config.h
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/aten_ops/aten_op_config.h:44]: not exactly sure what's the usage of backward_output_configs just by looking at the name, and the example below... 
maybe a comment to explain this?

File: orttraining/orttraining/python/training/ortmodule/_utils.py
Comment (SherlockNoMad) [inline orttraining/orttraining/python/training/ortmodule/_utils.py:148]: a comment to explain we apply TensorTransformFunc on input number 2.

File: orttraining/orttraining/python/training/ortmodule/_utils.py
Comment (SherlockNoMad) [inline orttraining/orttraining/python/training/ortmodule/_utils.py:153]: nit, reverse order?
input first, output last

File: orttraining/orttraining/python/training/ortmodule/_utils.py
Comment (SherlockNoMad) [inline orttraining/orttraining/python/training/ortmodule/_utils.py:169]: add a todo here, need to handle optional argument and arguments with default values.

File: orttraining/orttraining/python/training/ortmodule/_utils.py
Comment (SherlockNoMad) [inline orttraining/orttraining/python/training/ortmodule/_utils.py:156]: assert 
std::get<0>(raw_arguments[i]) < ivalue_arguments.size() ?

File: orttraining/orttraining/python/training/ortmodule/_utils.py
Comment (SherlockNoMad) [inline orttraining/orttraining/python/training/ortmodule/_utils.py:196]: think about how should we handle exception case here...

**Review by SherlockNoMad — CHANGES_REQUESTED** [review]
I went through this PR. We have a good framework to start with. 👍 
Left some comments, mostly are cosmetic/defensive programming/error handling... nothing fundamental needs to be changed... 
We should be be able to merge this after 1 more iteration of review.

File: orttraining/orttraining/core/graph/training_op_defs.cc
Comment (iK1D) [reply orttraining/orttraining/core/graph/training_op_defs.cc:2501]: We are using symbolic shape for ORTModule so we cannot support the shape inference here. I added some logic to symbolic_shape_infer.py to do this.

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py
Comment (iK1D) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:114]: I moved this call to after export and call it only when there is ATenOp in the graph.

File: orttraining/orttraining/training_ops/cuda/aten_ops/aten_op.cc
Comment (iK1D) [reply orttraining/orttraining/training_ops/cuda/aten_ops/aten_op.cc:12]: We uses SetOutputMLValue to set the OrtValue back, and SetOutputMLValue only allows setting external outputs. I am thinking external output means the output is from external (not allocated by ORT), whether the creator still hold the reference of the tensor is not important.

File: cmake/onnxruntime_providers.cmake
Comment (iK1D) [reply cmake/onnxruntime_providers.cmake:212]: We still need it as "#include "core/dlpack/dlpack_converter.h" will try to include <dlpack/dlpack.h> at the end. It won't compile if remove.

File: orttraining/orttraining/training_ops/cpu/aten_ops/aten_op_executor.h
Comment (iK1D) [reply orttraining/orttraining/training_ops/cpu/aten_ops/aten_op_executor.h:35]: I am thinking static instance is same as std::call_once, the problem is how we support singleton with argument in constructor. I added an Initialize static function to support that.

File: orttraining/orttraining/training_ops/cpu/aten_ops/aten_op_config.h
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/aten_ops/aten_op_config.h:94]: nit, reuse the iterator returned from .find()?

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py
Comment (SherlockNoMad) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py]: nice!

Comment (SherlockNoMad) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py]: OK

File: orttraining/orttraining/training_ops/cpu/aten_ops/aten_op_config.h
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/aten_ops/aten_op_config.h:53]: >backward_output_configs [](http://example.com/codeflow?start=19&length=23)

maybe find a better name for this ?

**Review by SherlockNoMad — DISMISSED** [review]

File: orttraining/orttraining/python/training/ortmodule/_utils.py
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/_utils.py:118]: Do we have to check twice, before and after the lock? How about locking first and then checking? would save one check

File: orttraining/orttraining/python/training/ortmodule/_utils.py
Comment (iK1D) [reply orttraining/orttraining/python/training/ortmodule/_utils.py:109]: Still keep the aten_op_executor, as I am thinking maybe the _load_torch_gpu_allocator_cpp_extension also need its run_once...

File: orttraining/orttraining/core/graph/training_op_defs.cc
Comment (iK1D) [reply orttraining/orttraining/core/graph/training_op_defs.cc:2567]: Could you help to have someone work on this as I am not quite familiar with exporter?

**Review by thiagocrepaldi — APPROVED** [review]

File: tools/python/register_custom_ops_pytorch_exporter.py
Comment (thiagocrepaldi) [inline tools/python/register_custom_ops_pytorch_exporter.py:38]: Is there any reason to not always register this op? Maybe we dont want that for inference? If that is the case, "is_training" would be more meaningfull than "is_ortmodule".

**Pull request merged**
